### PR TITLE
`_all` should be treated as an ordinary field when returning `fields` or `fielddata_fields`

### LINKED ISSUE #13178: fielddata_fields for _all doesn't display all the terms
Tested in 1.7.1.
For the following test

```
PUT /test_qs
{
  "mappings": {
    "test_type": {
      "properties": {
        "text1": {
          "type": "string"
        },
        "text2": {
          "type": "string"
        },
        "text3": {
          "type": "string"
        }
      }
    }
  }
}

POST /test_qs/test_type/_bulk
{"index":{}}
{"text1":"abc","text2":"def","text3":"ghi"}

GET /test_qs/test_type/_search
{
  "fielddata_fields": ["_all"]
}
```

the result is

```
            "fields": {
               "_all": "abc"
            }
```

where it should have been actually `"_all": ["abc", "def", "ghi"]`.

### PR DESCRIPTION
**Pull request by xuzha**
The '_all' meta field is not stored, but our fielddata field could work with not stored field. 

Our meta-fields typically contain a single value only, when `toXContent`for meta field, we only return the first value, but _all should be an exception there. This commit adds all values for `_all` field in the  `toXContent` method

closes #13178

### CODE REVIEW

**Comment by clintongormley** [comment]
@xuzha could you add some description to this PR please, and some documentation?

@jpountz could you review please?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java [hunk 1]
```diff
@@ -438,7 +438,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 if (field.values().isEmpty()) {
                     continue;
                 }
-                if (field.isMetadataField()) {
+                // "_all" should be treated as an ordinary field, issue 13178
+                if (field.isMetadataField() && !field.name().equals("_all")) {
```
Comment (andrestc) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: I think you could use the constant [Metadata.ALL](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java#L85) here.

Comment (xuzha) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: @andrestc  Thanks, agree

**Comment by xuzha** [comment]
@clintongormley  thanks, just update description. 

This PR is pretty straightforward, what kind of document would you like to improve, like suggest not use `_all` for `fielddata field` ?

**Comment by clintongormley** [comment]
@xuzha Because the `_all` field now behaves differently to the other meta-fields, I'd add a section to its documentation explaining how it is different and why https://www.elastic.co/guide/en/elasticsearch/reference/master/mapping-all-field.html

Actually, before we merge this PR, I think we need some further discussion about whether this is the right thing to do...  The `_all` field today can be stored, in which case it is stored as a single value which is returned as a single value at the top-level (like other meta-fields).

The only anomaly is the use of `?fielddata_fields=_all`, which is likely to contain multiple values... But loading fielddata for the `_all` field is a really really bad idea.  That said, it is probably still something we should allow.  Grrr, I'm really not sure what is the right thing to do here.

@jpountz what do you think?

**Comment by rjernst** [comment]
Can we just remove the ability to store/load `_all`? This seems silly to do, and if someone really wants to do it, they can do it with their own regular field? `_all` is meant to be the default for how full text search is done. It isn't meant to be used in other ways (and highlighting is not a real usecase, it doesn't make sense since `_all` is a catchall field and maintains zero information about which field values came from).

**Comment by jpountz** [comment]
I tend to agree with what @rjernst said. I was first a bit reluctant to remove highlighting support, but since highlighting is about locating matches in the original text, it indeed probably doesn't make much sense to highlight on `_all` since it loses information about which fields the snippets come from.

**Comment by clintongormley** [comment]
Had a chat to @nik9000 about highlighting and the `_all` field.  I think we should leave things as they are for now, and revisit this once we have a better story around highlighting.

**Comment by dakrone** [comment]
@clintongormley should this be closed and re-opened if/when appropriate since it's pretty out of date?

**Comment by clintongormley** [comment]
Sure

**Pull request closed without merging**